snow/engine/common/timer.go

```diff
@@ -3,12 +3,113 @@
 
 package common
 
-import "time"
+import (
+	"sync"
+	"time"
+)
 
-// Timer describes the standard interface for specifying a timeout
-type Timer interface {
+// PreemptionSignal signals when to preempt the pendingTimeout of the timeout handler.
+type PreemptionSignal struct {
+	activateOnce sync.Once
+	initOnce     sync.Once
+	signal       chan struct{}
+}
+
+func (ps *PreemptionSignal) init() {
+	ps.signal = make(chan struct{})
+}
+
+// Listen returns a read-only channel that is closed when Preempt() is invoked.
+func (ps *PreemptionSignal) Listen() <-chan struct{} {
+	ps.initOnce.Do(ps.init)
+	return ps.signal
+}
+
+// Preempt causes any past and future calls of Listen to return a closed channel.
+func (ps *PreemptionSignal) Preempt() {
+	ps.initOnce.Do(ps.init)
+	ps.activateOnce.Do(func() {
+		close(ps.signal)
+	})
+}
+
+// timeoutScheduler schedules timeouts to be dispatched in the future.
+// Only a single timeout can be pending to be scheduled at any given time.
+// Once a preemption signal is closed, all timeouts are immediately dispatched.
+type timeoutScheduler struct {
+	newTimer         func(duration time.Duration) *time.Timer
+	onTimeout        func()
+	preemptionSignal <-chan struct{}
+	pendingTimeout   chan struct{}
+}
+
+// NewTimeoutScheduler constructs a new timeout scheduler with the given function to be invoked upon a timeout,
+// unless the preemptionSignal is closed and in which case it invokes the function immediately.
+func NewTimeoutScheduler(onTimeout func(), preemptionSignal <-chan struct{}, newTimer func(duration time.Duration) *time.Timer) *timeoutScheduler {
+	pendingTimout := make(chan struct{}, 1)
+	pendingTimout <- struct{}{}
+	return &timeoutScheduler{
+		preemptionSignal: preemptionSignal,
+		newTimer:         newTimer,
+		onTimeout:        onTimeout,
+		pendingTimeout:   pendingTimout,
+	}
+}
+
+// RegisterTimeout fires the function the timeout scheduler is initialized with no later than the given timeout.
+func (th *timeoutScheduler) RegisterTimeout(d time.Duration) {
+	acquiredToken := th.acquirePendingTimeoutToken()
+	preempted := th.preempted()
+
+	if !preempted && !acquiredToken {
+		return
+	}
```

Comment:
Why is there special handling around the preemption here? Specifically, I'm not sure why we allow multiple goroutines to be scheduled if we have been preempted. I think just having:
```golang
if !th.acquirePendingTimeoutToken() {
	return
}
```
should be sufficient for our expected invariants.

If we did this, I think we could get rid of the `acquiredToken` argument in `scheduleTimeout` and remove the `preempted` helper entirely.

Reply:
Because the behavior of the previous code scheduled a new goroutine regardless. 
The trick in the bottom select doesn't work in most cases as it's very time reliant - it schedules a goroutine regardless, and then attempts to "cancel it" after the timeout. 

```
func (h *handler) RegisterTimeout(d time.Duration) {
	go func() {
		timer := time.NewTimer(d)
		defer timer.Stop()

		select {
		case <-timer.C:
		case <-h.preemptTimeouts:
		}

		// If there is already a timeout ready to fire - just drop the
		// additional timeout. This ensures that all goroutines that are spawned
		// here are able to close if the chain is shutdown.
		select {
		case h.timeouts <- struct{}{}:
		default:
		}
	}()
}
```

I agree though I should probably take advantage of the opportunity and not replicate a pattern if it does not add any value.